### DISCUSSION TITLE
Audio Streaming Server with Axum

### DISCUSSION THREAD

**Discussion by Tahinli** [tokio-rs/axum#2658] [Q&A] [answered]
### Summary

Hi, I must develop audio streaming service. I use [Dioxus](https://github.com/DioxusLabs/dioxus) for frontend and Axum as a backend. I can play streams with html audio in Dioxus but I don't know how to create my own audio stream with URL. I'm lack of knowledge to find out what to do and how to do. I just know that I have to do and I should be able to do with these awesome Dioxus and Axum.

How can I create audio stream with URL that html audio component can consume and play without problem.

### axum version

0.7.4

**Comment by Tahinli** [comment]
I did this

```rust
async fn stream() -> impl IntoResponse {
    println!("Stream");
    let header = [
        (header::CONTENT_TYPE, "audio/mpeg"),
        (header::CONTENT_LENGTH, ""),
        (header::TRANSFER_ENCODING, "chunked"),
        (header::CACHE_CONTROL, "no-cache, must-revalidate"),
        (header::PRAGMA, "no-cache"),

    ];
    let file = File::open("audios/audio.mp3").await.unwrap();
    let stream = ReaderStream::new(file);
    (header,Body::from_stream(stream))
}
```

and connected to router.
Problem is people can see total length and play the time that they want.
![image](https://github.com/tokio-rs/axum/assets/96421894/77025f69-5fbc-4fe6-a5a4-44f5b65281d3)
But i want it to make media time equal for every client and prevent them to reach future or previous data like this. I don't want to make client fetch all data from server at once. They should not see total length also.
![image](https://github.com/tokio-rs/axum/assets/96421894/fcb723c2-1f4c-476e-ba15-d125e36743e8)

How can I do like this ?

**Comment by adriandelgado** [comment, marked as answer]
~You can use [`ServeFile`](https://docs.rs/tower-http/latest/tower_http/services/struct.ServeFile.html). See: https://github.com/tokio-rs/axum/discussions/608#discussioncomment-7772294~ Edit: not relevant to the question

Take a look at this PR: https://github.com/tokio-rs/axum/pull/3047/files#diff-85ae2acd48337869f0504aa4eced18f7129187436fedaeee5e5c941c41d6292d. Specifically at the `try_stream` function.
Also, take a look at how `ServeDir` implements sending a portion of a file https://docs.rs/tower-http/latest/src/tower_http/services/fs/serve_dir/future.rs.html#253. It seems like you want to manually send what part of the file you want. I recommend using an AtomicUsize to keep track of which part of the file you should send to everyone.

**Reply by Tahinli** [reply]
Hi, Does this one solve client sync for everybody ? I was trying to sync stream equal for everyone. I don't want to let people seek further or past about media but allow them to follow same moment of stream.

**Reply by adriandelgado** [reply]
I updated my answer. I didn't correctly read you question originally, sorry.

**Reply by Tahinli** [reply]
Thanks a lot.
